torchci/lib/searchLogs.ts

```diff
@@ -0,0 +1,78 @@
+import { JobData } from "./types";
+
+export async function getSearchRes(
+  jobs: JobData[],
+  query: string,
+  setSearchRes: any
+) {
+  // Helper function to be called in useEffect to make sure the component
+  // reloads when it should (hopefully).  The actual searching is done in
+  // searchLogs
+  if (query == "") {
+    setSearchRes({
+      results: new Map(),
+      info: undefined,
+    });
+    return;
+  }
+  setSearchRes({
+    results: new Map(),
+    info: "Loading... (this might take a while)",
+  });
+  const searchRes = await searchLogs(
+    jobs.map((job) => job.id!),
+    query
+  );
+  setSearchRes({
+    results: searchRes,
+    info: undefined,
+  });
+}
+
+async function searchLogs(
+  jobIds: string[],
+  query: string
+): Promise<Map<string, [number[], string[]] | undefined>> {
+  if (query == "") {
+    return new Map();
+  }
+  const results: [string, [number[], string[]] | undefined][] =
+    await Promise.all(
+      jobIds.map(async (jobId) => {
+        return [jobId, await searchLog(jobId, query)];
+      })
+    );
+  return new Map(results);
+}
+
+async function searchLog(
+  jobId: string,
+  query: string
+): Promise<[number[], string[]] | undefined> {
+  // Search indiividual log
+  try {
+    const result = await fetch(
+      `https://ossci-raw-job-status.s3.amazonaws.com/log/${jobId}`
+    );
+    if (result.status != 200) {
+      return undefined;
+    }
+    const lineNumbers = [];
+    const lineTexts = [];
+    const threshold = 100;
+    for (const [index, line] of (await result.text()).split("\n").entries()) {
```

Comment:
We could also perform the search from the end of the logs here instead from the beginning because more relevant logs usually are at the end?

Reply:
I'm worried that will be confusing, also if someone wants the last 100 results, they can use the log viewer to search afterwards